hadoop-hdds/docs/content/design/s3_hcfs.md

```diff
@@ -0,0 +1,282 @@
+ * **aws s3 ls**: prefix (`PRE h/`) and file (`h`) are both displayed
+ * S3A: both entries are visible with the name `/a/b/h` but firt is a file (with size) second is a directory (with directory attributes)
+ 
+Proposed behavior:
+
+ * `ozone.om.enable.intermediate.dirs=true`: show both the file and the directory with the same name (similar to S3A)
+ * `ozone.om.enable.filesystem.paths=true`: throwing exception when the second one is created  
+
+## Directory entry created with file content
+
+In this case we create a directory (key which ends with `/`) but with real file content:
+
+```
+aws s3api put-object --bucket ozonetest --key a/b/i/ --body README.md
+```
+
+Behavior:
+ * *S3 web console*: rendered as directory (couldn't be downloaded)
+ * **aws s3 ls**: showed as a prefix (`aws s3 ls s3://ozonetest/a/b`), but when the full path is used showed as a file without name (`aws s3 ls s3://ozonetest/a/b/i/`)
+ * S3A: `./bin/hdfs dfs -ls s3a://ozonetest/a/b/` shows a directory `h`, `./bin/hdfs dfs -ls s3a://ozonetest/a/b/i` shows a file `i`
+ 
+Proposed behavior:
+
+ * `ozone.om.enable.intermediate.dirs=true`: possible but `i/` is hidden from o3fs/ofs
+ * `ozone.om.enable.filesystem.paths=true`: key name is normalized to real key name
+
+## Create key and explicit create parent dir
+
+```
+aws s3api put-object --bucket ozonetest --key e/f/g/
+aws s3api put-object --bucket ozonetest --key e/f/
+```
+
+Behavior:
+
+ * S3 can support it without any problem
+ 
+Proposed behavior:
+
+After the first command `/e/f/` and `/e/` entries created in the key space (as they are required by `ofs`/`o3fs`) but **with a specific flag** (explicit=false). 
+
+AWS S3 list-objects API should exclude those entries from the result (!).
+
+Second command execution should modify the flag of `/e/f/` key to (explicit=true).
+
+## Create parent dir AND key with S3a
+
+This is the problem which is reported by [HDDS-4209](https://issues.apache.org/jira/browse/HDDS-4209)
+
+```
+hdfs dfs -mkdir -p s3a://b12345/d11/d12 # -> Success
+
+hdfs dfs -put /tmp/file1 s3a://b12345/d11/d12/file1 # -> fails with below error
+```
+
+Proposed behavior:
+
+ * `ozone.om.enable.intermediate.dirs=true`: should work without error
+ * `ozone.om.enable.filesystem.paths=true`: should work without error.
+
+This is an `ofs`/`o3fs` question not an S3. The directory created in the first step shouldn't block the creation of the file. This can be a **mandatory** normalization for `mkdir` directory creation. As it's an HCFS operation, s3 is not affected. Entries created from S3 can be visible from s3 without any problem.
+
+## Create file and directory with S3
+
+This problem is reported in HDDS-4209, thanks to @Bharat
+
+```
+hdfs dfs -mkdir -p s3a://b12345/d11/d12 -> Success
+
+hdfs dfs -put /tmp/file1 s3a://b12345/d11/d12/file1 
+```
+
+In this case first a `d11/d12/` key is created. The intermediate key creation logic in the second step should use it as a directory instead of throwing an exception.
```

Comment:
d11/d12 is created without traiiling "/"

Reply:
I tested it locally and found the trailing `/`. Are you sure?

Reply:
Yes. Posted my comment in HDDS-4209. Refer for more info

Reply:
Thanks to explain it. I tested it with pure AWS S3 and `s3a`. The trailing `/` seems to be missing due to a bug in our normalization, and it's not and `s3a` behavior. I think we should fix the normalization.